### PR TITLE
dmd.mtype: Move modifiersApply and attributesApply to free functions

### PR DESCRIPTION
**Pull request by ibuclaw**
Remove `extern(C++)` linkage, and replace `const(char)*` with `string`

### CODE REVIEW

File: src/dmd/mtype.d [hunk 1]
```diff
@@ -6521,3 +6459,80 @@ const(char*)[2] toAutoQualChars(Type t1, Type t2)
     }
     return [s1, s2];
 }
+
+
+/**
+ * For each active modifier (MODFlags.const_, MODFlags.immutable_, etc) call `fp` with a
+ * void* for the work param and a string representation of the attribute.
+ */
+final int modifiersApply(TypeFunction tf, void* param, int function(void*, const(char)[]) fp)
```
Comment (jacob-carlborg) [inline src/dmd/mtype.d]: * Look at this on my phone, if this is a free function it should not be `final`
* Missing standard Ddoc sections

File: src/dmd/mtype.d [hunk 2]
```diff
@@ -6521,3 +6459,80 @@ const(char*)[2] toAutoQualChars(Type t1, Type t2)
...
+        }
+    }
+
+    return 0;
+}
+
+/**
+ * For each active attribute (ref/const/nogc/etc) call `fp` with a void* for the
+ * work param and a string representation of the attribute.
+ */
```
Comment (jacob-carlborg) [inline src/dmd/mtype.d]: Same here, missing standard Ddoc sections.

**Review by RazvanN7 — APPROVED** [review]

File: src/dmd/mtype.d [hunk 3]
```diff
@@ -6521,3 +6459,80 @@ const(char*)[2] toAutoQualChars(Type t1, Type t2)
...
+int modifiersApply(TypeFunction tf, void* param, int function(void*, const(char)[]) fp)
+{
+    immutable ubyte[4] modsArr = [MODFlags.const_, MODFlags.immutable_, MODFlags.wild, MODFlags.shared_];
+
+    foreach (modsarr; modsArr)
+    {
+        if (tf.mod & modsarr)
+        {
+            if (int res = fp(param, MODtoString(modsarr)))
```
Comment (WalterBright) [inline src/dmd/mtype.d]: This is a poor solution (even though what you're replacing did the same thing) in that it is pointlessly allocating memory to make a copy of a string literal, and leaking that memory.

File: src/dmd/mtype.d [hunk 4]
```diff
@@ -214,6 +214,16 @@ char* MODtoChars(MOD mod) nothrow
...
 }
 
+/// Ditto
+const(char)[] MODtoString(MOD mod) nothrow
+{
+    OutBuffer buf;
+    buf.reserve(16);
+    MODtoBuffer(&buf, mod);
+    size_t len = buf.offset;
+    return buf.extractData()[0 .. len];
```
Comment (WalterBright) [inline src/dmd/mtype.d]: This implementation should be swapped with the implementation of ModtoBuffer, then there will be no need to make the copy.

**Review by WalterBright — CHANGES_REQUESTED** [review]
Might as well fix the underlying issue it has with leaking memory.

File: src/dmd/mtype.d (hunk 3 above)
Comment (ibuclaw) [reply src/dmd/mtype.d]: Call ModToBuffer from here? OK, seems reasonable.

File: src/dmd/mtype.d [hunk 5]
```diff
@@ -6521,3 +6449,83 @@ const(char*)[2] toAutoQualChars(Type t1, Type t2)
...
+    immutable ubyte[4] modsArr = [MODFlags.const_, MODFlags.immutable_, MODFlags.wild, MODFlags.shared_];
+
+    foreach (modsarr; modsArr)
+    {
+        if (tf.mod & modsarr)
+        {
+            OutBuffer buf;
+            MODtoBuffer(&buf, modsarr);
+            const len = buf.offset;
+            if (int res = fp(param, buf.extractData()[0 .. len]))
```
Comment (WalterBright) [inline src/dmd/mtype.d]: It still leaks memory, once each time through the loop. The strings produced by MODtoString are all literals and don't need to be duplicated.

Comment (WalterBright) [reply src/dmd/mtype.d]: I mean MODtoBuffer.

**Comment by ibuclaw** [comment]
MODtoString returns a constant.

**Comment by ibuclaw** [comment]
Ping?

**Comment by RazvanN7** [comment]
ping @WalterBright

**Review by WalterBright — APPROVED** [review]

**Pull request merged**